assets/shared/helpers/player/index.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Internal dependencies
+ */
+import * as videoFileAdapter from './video-file-adapter';
+import * as videopressAdapter from './videopress-adapter';
+import * as youtubeAdapter from './youtube-adapter';
+import * as vimeoAdapter from './vimeo-adapter';
+
+const VIDEO_TYPE = videoFileAdapter.ADAPTER_NAME;
+const VIDEOPRESS_TYPE = videopressAdapter.ADAPTER_NAME;
+const YOUTUBE_TYPE = youtubeAdapter.ADAPTER_NAME;
+const VIMEO_TYPE = vimeoAdapter.ADAPTER_NAME;
+
+const adapters = {
+	[ VIDEO_TYPE ]: videoFileAdapter,
+	[ VIDEOPRESS_TYPE ]: videopressAdapter,
+	[ YOUTUBE_TYPE ]: youtubeAdapter,
+	[ VIMEO_TYPE ]: vimeoAdapter,
+};
+
+/**
+ * A class that abstracts the use of the player APIs: Video, VideoPress, YouTube, and Vimeo.
+ */
+class Player {
+	/**
+	 * Player constructor.
+	 *
+	 * @param {HTMLVideoElement|HTMLIFrameElement} element The player element.
+	 * @param {Window}                             w       A custom window.
+	 */
+	constructor( element, w = window ) {
+		this.playerPromise = null;
+		this.adapterName = null;
+		this.element = element;
+		this.w = w;
+
+		this.setAdapter();
+	}
+
+	/**
+	 * Set the player adapter.
+	 */
+	setAdapter() {
+		if ( this.element instanceof this.w.HTMLVideoElement ) {
+			this.adapterName = VIDEO_TYPE;
+		} else if ( this.element instanceof this.w.HTMLIFrameElement ) {
+			this.adapterName = Object.entries( adapters ).find(
+				( [ , { EMBED_PATTERN = null } ] ) =>
+					EMBED_PATTERN && this.element.src?.match( EMBED_PATTERN )
+			)?.[ 0 ];
+		}
+
+		if ( ! this.adapterName ) {
+			// eslint-disable-next-line no-console -- We want to expose the element with problem.
+			console.error( 'Video adapter not found', this.element );
+		}
+	}
+
+	/**
+	 * Get the video player.
+	 *
+	 * @return {Promise<Object|HTMLVideoElement|HTMLIFrameElement>} The video player through a promise.
+	 */
+	getPlayer() {
+		if ( ! this.playerPromise ) {
+			this.playerPromise =
+				adapters[ this.adapterName ]?.initializePlayer(
```

Comment:
Here we are using the `?` operator in case the adapter wasn't set, but I notice that in the other methods (`getDuration`, `setCurrentTime`, `play`, `pause`, and `on`) we don't do so. Should we be doing so? I wonder if it would be better to have a method `getAdapter` which does the check and throws an exception if it doesn't exist? Maybe it would be even better to throw an exception from the constructor if the adapter wasn't created properly?

Reply:
> Should we be doing so?

Not necessarily, because all the methods call the `getPlayer` internally, and if `getPlayer` fails, it will never be resolved in the other methods because the `Promise.reject( new Error( 'Failed getting the player' ) );`.

> Maybe it would be even better to throw an exception from the constructor if the adapter wasn't created properly?

I thought about that and I had some refactors around that. Currently, the constructor is calling the `setAdapter` that will just call a `console.error( 'Video adapter not found', this.element );` in case of error. I didn't use the throw to avoid breaking the rest of the code that would be calling this, in case it doesn't handle the error. But if you really prefer, we could replace the `console.error` with a `throw`, and assume that it should be properly handled when calling this.

Reply:
> Not necessarily, because all the methods call the getPlayer internally, and if getPlayer fails, it will never be resolved in the other methods

Ahhh thanks, I missed that detail. Makes sense 👍 

It seems like throwing an error from the constructor might make this more robust, but I see what you mean about breaking stuff too. I think it's fine either way.

Reply:
Anyway, I created the `getAdapter` so the methods don't need to be accessing the attribute directly, and getting it from the specific key every time. It seems better! Thank you for the suggestion!